### DISCUSSION TITLE
About the workflow definition parameter tenantid

### DISCUSSION THREAD

**Discussion by GreenShadeZhang** [elsa-workflows/elsa-core#1197] [Q&A]
## When I look at the interface document and source code, I find that tenantid cannot be passed in when creating workflow definition. I want to ask whether this operation is not supported or the design is so.

**Comment by GreenShadeZhang** [comment]
## Below is a screenshot of the code
![image](https://user-images.githubusercontent.com/24240675/124100782-540a3400-da91-11eb-8780-1cc6bc851a17.png)

**Comment by sfmskywalker** [comment]
Although the TenantId field is there, multi-tenancy hasn't been implemented yet, so it is indeed not yet supported. It's unclear at this point when I can work on it, but we have a couple of issues related to this topic: #242 and #762.

**Reply by GreenShadeZhang** [reply]
> Although the TenantId field is there, multi-tenancy hasn't been implemented yet, so it is indeed not yet supported. It's unclear at this point when I can work on it, but we have a couple of issues related to this topic: #242 and #762.

I found that when I query data, I can pass in tenantid and filter it normally. Is it possible for me to implement the API to work normally? I just need to simply identify the data including tenantid. Of course, I may be a custom dashboard.

**Reply by sfmskywalker** [reply]
Sure, if you want you can create a PR which I’ll gladly accept.

**Reply by GreenShadeZhang** [reply]
> Sure, if you want you can create a PR which I’ll gladly accept.

The scenario I can think of is that tenantid is obtained from JWT token or other authentication methods, so it may not be easy to pass in directly from API parameters. It should be from httpcontext. User to get claim. There is a key called tenantid. But this feeling will affect your API project. I see that some packages of some projects will introduce some < packagereference include = "Microsoft. Aspnetcore" version = "2.2.0" / >
< packagereference include = "Microsoft. Aspnetcore. MVC" version = "2.2.0" / >, I feel you can replace them with < frameworkreference include = "Microsoft. Aspnetcore. App" / >

**Reply by sfmskywalker** [reply]
What if we introduce an abstraction? For example, an interface called `ITenantAccessor`. Then we can provide a number of concrete implementations from separate packages. One package would provide a couple of implementations that read the tenant from the HTTP request, like current domain, path prefix or header. Another would read from user claims.

Best of all, the developer can implement their own strategy that is specific to their use case.

**Reply by GreenShadeZhang** [reply]
> What if we introduce an abstraction? For example, an interface called `ITenantAccessor`. Then we can provide a number of concrete implementations from separate packages. One package would provide a couple of implementations that read the tenant from the HTTP request, like current domain, path prefix or header. Another would read from user claims.
> 
> Best of all, the developer can implement their own strategy that is specific to their use case.

This sounds good, so that we can adopt the default behavior in the API, and then when we inject the custom behavior into the startup, we can provide tenantid. We can provide tenantid in creating workflow definition in various ways. Or we can use this provider when querying.

**Reply by GreenShadeZhang** [reply]
Take the liberty to ask, when can you add this method to the source code.

**Reply by sfmskywalker** [reply]
Hard to say at this point. Right now my main priority is getting 2.1 to be released and work on a number of requirements for a customer. Multi-tenancy is a big task, so it depends on multiple factors when I might work on this. I was kinda hoping you would do it :)

**Reply by GreenShadeZhang** [reply]
> Hard to say at this point. Right now my main priority is getting 2.1 to be released and work on a number of requirements for a customer. Multi-tenancy is a big task, so it depends on multiple factors when I might work on this. I was kinda hoping you would do it :)

I have made some basic adjustments according to the methods we discussed two days ago and the contents you discussed before. These contents are mainly used in elsa.server.api. If you think it is necessary to modify them, please let me know. I have proposed a pr. you can see if it is feasible. I have also provided a test example, The API swagger of the master branch will report an error. I changed the API version of Getmany before swagger works normally. Therefore, you may need to see if the API name conflicts.

![image](https://user-images.githubusercontent.com/24240675/124781631-23853700-df76-11eb-876f-6ad85d92d3e1.png)
